### PR TITLE
Separate the part of the nametab state related to module names

### PR DESCRIPTION
**Pull request by maximedenes**
This adds a bit more structure to the nametab state, while preserving backward compatibility. Needed for #15409.

### CODE REVIEW

File: library/nametab.ml
Comment (ppedrot) [inline library/nametab.ml:586]: Is it possible to group all this stuff in a single `ref`? The use of `Summary.ref` is clearly indicating a pattern in the code while here it's a bit drowned in the noise. (This is not unheard of, but ideally we should stop manipulating the state by hand and only rely on the combinators if we want to move to a finer implementation.)

**Comment by ppedrot** [comment]
I am assigning because it seems that there are not many library maintainers.

File: library/nametab.ml
Comment (maximedenes) [reply library/nametab.ml:586]: Sure! Do you want a record with mutable fields, or a ref with functional updates?

Comment (ppedrot) [reply library/nametab.ml:586]: The latter.

Comment (maximedenes) [reply library/nametab.ml]: Here you go!

**Review by ppedrot — APPROVED** [review]

File: library/nametab.ml
Comment (silene) [inline library/nametab.ml:393]: Just a minor nitpick in case you need to modify the pull request before it is merged. Never use the idiom above, as it confuses OCaml to no end. It causes it to allocate a closure (and then to resolve it when called). Always eta-expand module values, i.e., `let unfreeze_modules_nametab x = modules_nametab := x`.

(Perhaps Flambda knows how to compile this idiom correctly, but since we are not using Flambda, this does not matter.)

Comment (ppedrot) [reply library/nametab.ml:393]: I agree that all functions should always be eta-expanded in OCaml. It's easy to introduce bugs with partial applications. (Here it's innocuous, but see e.g. #15726 for a nasty example.)

Comment (maximedenes) [reply library/nametab.ml:393]: Sure, I can fix it. Grepping the codebase seems to show quite a few occurrences, btw.

Comment (maximedenes) [reply library/nametab.ml]: Fixed.

File: library/nametab.mli
Comment (ejgallego) [inline library/nametab.mli:251]: Just a nit on style, maybe it'd be more convenient to use a module here.
```
module Modules : sig
  type t
  val freeze ...
  val unfreezze
end
```
At some point we may want to have a "Freezable" interface.

File: library/nametab.ml
Comment (ejgallego) [inline library/nametab.ml:379]: Similar comment on maybe using `Modules.t` ?

Comment (maximedenes) [reply library/nametab.ml:379]: Sure, how would you call the `modules_nametab` ref, in this style?

Comment (maximedenes) [reply library/nametab.ml]: It will be `Modules.nametab`.

Comment (maximedenes) [reply library/nametab.ml]: Done.

File: library/nametab.mli
Comment (maximedenes) [reply library/nametab.mli]: Done.

File: library/nametab.ml
Comment (Alizter) [reply library/nametab.ml]: @maximedenes Can you paste your grep results?

**Review by ejgallego — APPROVED** [review]
Looks pretty good to me @ppedrot

File: library/nametab.ml
Comment (maximedenes) [reply library/nametab.ml]: Ok, but that's for a (potential) separate patch:

```
lib/flags.ml:    let () = List.iter2 (:=) ol vl in r
lib/flags.ml:    let () = List.iter2 (:=) ol vl in
lib/flags.ml:let set_inline_level = (:=) inline_level
library/summary.ml:      unfreeze_function = ((:=) r);
plugins/extraction/common.ml:  let ph = ref Impl in ((:=) ph), (fun () -> !ph)
plugins/extraction/common.ml:  ((:=) k), (fun () -> !k)
plugins/ltac/tauto.ml:      optwrite = (:=) negation_unfolding }
pretyping/detyping.ml:      optwrite = (:=) print_evar_arguments }
pretyping/program.ml:let set_proofs_transparency = (:=) proofs_transparency
pretyping/program.ml:    optwrite = (:=) program_cases }
pretyping/program.ml:    optwrite = (:=) program_generalized_coercion }
printing/ppconstr.ml:  let set_term_pr = (:=) term_pr
tactics/auto.ml:      optwrite = (:=) refe })
tactics/class_tactics.ml:  let set_typeclasses_debug d = (:=) typeclasses_debug (if d then 1 else 0)
tactics/eauto.ml:      optwrite = (:=) global_debug_eauto })
tactics/eauto.ml:      optwrite = (:=) global_info_eauto })
vernac/attributes.ml:        optwrite = ((:=) b) }
vernac/comSearch.ml:      optwrite = (:=) search_output_name_only }
vernac/prettyp.ml:let set_object_pr = (:=) object_pr
vernac/printmod.ml:      optwrite = ((:=) short) }
vernac/proof_using.ml:      optwrite = ((:=) suggest_proof_using) })
vernac/vernacentries.ml:      optwrite = ((:=) Flags.quiet) }
vernac/vernacentries.ml:      optwrite = (:=) Vmbytegen.dump_bytecode }
vernac/vernacentries.ml:      optwrite = (:=) Vmlambda.dump_lambda }
vernac/vernacinterp.ml:      optwrite = ((:=) default_timeout) }
```

**Comment by ppedrot** [comment]
@coqbot merge now

**Comment by maximedenes** [comment]
Thanks!

**Pull request merged**
